File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    int dfs(TreeNode* root, int &maxSum) {
        if(root == NULL)
            return 0;
        int lmax = dfs(root->left, maxSum);
        int rmax = dfs(root->right, maxSum);
        maxSum = max({lmax+root->val, rmax+root->val, rmax+lmax+root->val, maxSum, root->val});
        return max(root->val+max(lmax, rmax), root->val);
    }
public:
    int maxPathSum(TreeNode* root) {
        int maxSum = INT_MIN;
        dfs(root, maxSum);
        return maxSum;
    }
};

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    
    private int maxHelper(TreeNode root, ref int maxPathSum) {
        if(root == null)
            return 0;
        int lh = maxHelper(root.left,ref maxPathSum);
        int rh = maxHelper(root.right,ref maxPathSum);
        maxPathSum = new int[] { root.val+lh, root.val+rh, root.val+lh+rh, root.val, maxPathSum }.Max();
        return Math.Max(root.val+Math.Max(lh,rh), root.val);
    }
    
    public int MaxPathSum(TreeNode root) {
        int maxPathSum = root.val;
        maxHelper(root,ref maxPathSum);
        return maxPathSum;
    }
}